server/filestore.go

```diff
@@ -5295,6 +5313,45 @@ func (fs *fileStore) cancelAgeChk() {
 	}
 }
 
+// Lock must be held so that nothing else can interleave and write a
+// new message on this subject before we get the chance to write the
+// system tombstone. If the tombstone is written successfully then
+// this function returns a callback func to call scb and tcb after the
+// lock has been released.
+func (fs *fileStore) systemTombstoneIfNeeded(sm *StoreMsg, reason string) func() {
+	// If the deleted message was itself a tombstone then don't
+	// write out more tombstones or we'll churn endlessly.
+	if len(getHeader(JSAppliedLimit, sm.hdr)) != 0 {
+		return nil
+	}
+	if !fs.cfg.AllowMsgTTL || fs.cfg.LimitsTTL <= 0 {
+		return nil
+	}
+	if _, ok := fs.psim.Find(stringToBytes(sm.subj)); ok {
+		// There are still messages left with this subject,
+		// therefore it wasn't the last message deleted.
+		return nil
+	}
+	// Build the system tombstone.
+	var _hdr [128]byte
+	hdr := fmt.Appendf(_hdr[:0], "NATS/1.0\r\n%s: %s\r\n%s: %s\r\n\r\n", JSAppliedLimit, reason, JSMessageTTL, fs.cfg.LimitsTTL)
```

Comment:
LimitsTTL seems like wrong name IMO.

Reply:
Best that @ripieneer has input here as this is the name from the ADR.

Reply:
Best that @ripienaar has input here as this is the name from the ADR.

Reply:
When limits are applied and tombstones are written, this is the TTL of those messages.

We didnt consider purges would be involved until later, do you have a name suggestion @derekcollison ?

Reply:
Naming is hard, so not at the moment but will give it some thought.

Reply:
something like `ActionTTL` or `NotificationTTL`? Other ideas? Failing to find a better name :)

Reply:
Is it weird that the functionality is only turned on with a TTL? Do we want something like `LimitsTombstones bool` and `LimitsTombstonesTTL time.Duration`? What if you want tombstones but you want them to "live forever"?

Reply:
could set LimitsTTL to `never`? Just seemed excessive to add 2 fields

Reply:
It feels weird still, if it's `LimitsTombstones` then `never` doesn't make sense for something that *does* leave TTLs, and if it's `LimitsTTL` or `LimitsTombstonesTTL` then it's difficult to explain why the option is named after TTLs when it has a secondary larger effect.

Reply:
We can add 2 fields, I am not super stuck to the idea or anything, just conscious of how many config options we have, though I agree making it awkward is also not good

Reply:
I think the turning on should be a bool that is clear we will generate markers when a subject disappears, and that should have a default TTL. We can have a second that allows them to override or change the TTL for he markers, including a "never" if needed.

Reply:
How about `tombstoneTTL`?

Reply:
We've gone with `SubjectDeleteMarkers bool` and `SubjectDeleteMarkerTTL string`.

The default if not specified will be 15 minutes if not specified, RI and I think this is a good value as it's well outside the recovery time for fetch/ordered consumer/client/route/gateway etc failures as well as even application restarts.